pmlc/rt/runtime_registry.cc

```diff
@@ -2,89 +2,31 @@
 
 #include "pmlc/rt/runtime_registry.h"
 
-#include <list>
-#include <mutex>
-#include <stdexcept>
-#include <utility>
-
-#include "pmlc/util/logging.h"
-#include "llvm/Support/FormatVariadic.h"
+#include <string>
+#include <unordered_map>
 
 namespace pmlc::rt {
-namespace {
-
-struct LoaderList {
-  std::mutex mutex;
-  std::list<std::pair<std::string, Loader>> loaders;
-
-  static LoaderList &instance() {
-    static LoaderList ll;
-    return ll;
-  }
-};
-
-class MemoizedFactory final {
-public:
-  explicit MemoizedFactory(Factory factory) : factory{std::move(factory)} {}
-  Runtime *getRuntime() {
-    std::call_once(init_once, [this]() { runtime = factory(); });
-    return runtime.get();
-  }
-
-private:
-  std::once_flag init_once;
-  Factory factory;
-  std::shared_ptr<Runtime> runtime;
-};
-
-const std::unordered_map<std::string, std::function<Runtime *()>>
-buildRuntimeMap() {
-  std::unordered_map<std::string, std::function<Runtime *()>> result;
-
-  auto &ll = LoaderList::instance();
-  std::lock_guard<std::mutex> lock{ll.mutex};
 
-  for (auto &[loaderId, loader] : ll.loaders) {
-    std::unordered_map<std::string, Factory> factories;
-    try {
-      factories = loader();
-    } catch (const std::exception &e) {
-      IVLOG(1, "Loader " << loaderId << " initialization failed: " << e.what());
-      continue;
-    }
-    for (auto &[id, factory] : factories) {
-      auto memo = std::make_shared<MemoizedFactory>(factory);
-      auto [it, inserted] =
-          result.emplace(id, [memo = std::move(memo)]() -> Runtime * {
-            return memo->getRuntime();
-          });
-      if (!inserted) {
-        throw std::runtime_error{
-            llvm::formatv("Multiple runtime implementations found for {}", id)};
-      }
-    }
-  }
-
-  return result;
+std::unordered_map<std::string, std::unique_ptr<Runtime>> &getRuntimeMap() {
+  static std::unordered_map<std::string, std::unique_ptr<Runtime>> runtimeMap;
+  return runtimeMap;
 }
 
-} // namespace
-
 namespace details {
 
-void registerLoader(llvm::StringRef id, Loader loader) {
-  auto &ll = LoaderList::instance();
-  std::lock_guard<std::mutex> lock{ll.mutex};
-  ll.loaders.emplace_back(id, std::move(loader));
+void registerRuntime(llvm::StringRef id, std::unique_ptr<Runtime> runtime) {
+  getRuntimeMap()[id.str()] = std::move(runtime);
 }
 
 } // namespace details
 
-const std::unordered_map<std::string, std::function<Runtime *()>> &
-getRuntimeMap() {
-  static std::unordered_map<std::string, std::function<Runtime *()>>
-      runtimeMap = buildRuntimeMap();
-  return runtimeMap;
+void initRuntimes() {
+  [[maybe_unused]] static bool inited = []() {
```

Comment:
I've been using 
```
    static std::once_flag is_initialized;
    std::call_once(is_initialized, []() {
      // do init here
    });
```

Any reason why we shouldn't use that here?

To me, I can look at the `std::call_once` docs to understand how it works, while looking at a lambda in c++ with a strange attribute makes me pause and wonder if it does the right thing or not (I think it does...).

Reply:
They probably compile to the same code.

The static bool is guaranteed to be initialized exactly once, the first time the block is entered, with a barrier to synchronize parallel callers, and re-trying if there's an exception -- this is why Meyers singletons work.

With `call_once`: a naive compiler might generate code that checks that `is_initialized` has been initialized every single time the function is entered, and then do another atomic check + barrier in `call_once` to see whether it should run the lambda. But a smart compiler can use `once_flag`'s constexpr constructor to do the initialization at compilation time and elide the first runtime check, so I think it's fine.

Reviewing it myself, though, I think there's a bigger problem: I'm not sure what to do when a `Runtime::init()` throws an exception; the caller might call `initRuntimes()` again, so we either need to make `Runtime::init()` idempotent or we need to remember that we've called a Runtime's `init()` method somewhere.  And we should probably continue on to initialize all of the `Runtime`s, or declare that you can't use the system unless all `Runtime`s successfully initialize and `initRuntimes()` completes normally.  I'll work on this a bit more and send an update.

Reply:
I think Franks point is the call_once is easier to read (and I agree), and even if the compiler is dumb (hint: it's not), we save < 1us per call to initRuntimes, which is probably called exactly once anyway, so use the easier to read one.

Reply:
Sorry -- forgot to add that I'm happy to change it if you guys prefer `call_once`.
Should we also do the same for other cases where we use Meyers singletons, e.g. `getRuntimeMap()`?

Reply:
Meyers singletons are reasonable, it's the usage of a lambda with the non-standard attribute + being a Meyers singleton that gives me pause.

I'd prefer if we used `std::call_once` in the places that it makes sense (to me the impl of `getRuntimeMap` does not qualify). If we can encapsulate the initialization in the construction of an object, then I'd say a Meyers singleton is great. But in this case, we're calling a sequence of code, that seems like better suited for `std::call_once`

Reply:
SGTM.  (FWIW, `maybe_unused` is a standard attribute.  :-)

Reply:
Done

Reply:
TIL